source/extensions/common/wasm/wasm_vm.h

```diff
@@ -0,0 +1,309 @@
+   * @param needs_emscripten whether emscripten support should be provided (e.g.
+   * _emscripten_memcpy_bigHandler). Emscripten (http://https://emscripten.org/) is
+   * a C++ WebAssembly tool chain.
+   */
+  virtual void link(absl::string_view debug_name, bool needs_emscripten) PURE;
+
+  /**
+   * Set memory layout (start of dynamic heap base, etc.) in the VM.
+   * @param stack_base the location in VM memory of the stack.
+   * @param heap_base the location in VM memory of the heap.
+   * @param heap_base_ptr the location in VM memory of a location to store the heap pointer.
+   */
+  virtual void setMemoryLayout(uint64_t stack_base, uint64_t heap_base,
+                               uint64_t heap_base_pointer) PURE;
+
+  /**
+   * Call the 'start' function and initialize globals.
+   * @param vm_context a context which represents the caller: in this case Envoy itself.
+   */
+  virtual void start(Context* vm_context) PURE;
+
+  /**
+   * Get size of the currently allocated memory in the VM.
+   * @return the size of memory in bytes.
+   */
+  virtual uint64_t getMemorySize() PURE;
+
+  /**
+   * Convert a block of memory in the VM to a string_view.
+   * @param pointer the offset into VM memory of the requested VM memory block.
+   * @param size the size of the requested VM memory block.
+   * @return if std::nullopt then the pointer/size pair were invalid, otherwise returns
+   * a host string_view pointing to the pointer/size pair in VM memory.
+   */
+  virtual absl::optional<absl::string_view> getMemory(uint64_t pointer, uint64_t size) PURE;
+
+  /**
+   * Convert a host pointer to memory in the VM into a VM "pointer" (an offset into the Memory).
+   * @param host_pointer a pointer to host memory to be converted into a VM offset (pointer).
+   * @param vm_pointer a pointer to an uint64_t to be filled with the offset in VM memory
+   * corresponding to 'host_pointer'.
+   * @return whether or not the host_pointer was a valid VM memory offset.
+   */
+  virtual bool getMemoryOffset(void* host_pointer, uint64_t* vm_pointer) PURE;
+
+  /**
+   * Set a block of memory in the VM, returns true on success, false if the pointer/size is invalid.
+   * @param pointer the offset into VM memory describing the start of a region of VM memory.
+   * @param size the size of the region of VM memory.
+   * @return whether or not the pointer/size pair was a valid VM memory block.
+   */
+  virtual bool setMemory(uint64_t pointer, uint64_t size, const void* data) PURE;
+
+  /**
+   * Get a Word in the VM, returns true on success, false if the pointer is invalid.
+   * @param pointer the offset into VM memory describing the start of VM native word size block.
+   * @param data a pointer to a Word whose contents will be filled from the VM native word at
+   * 'pointer'.
+   * @return whether or not the pointer was to a valid VM memory block of VM native word size.
+   */
+  virtual bool getWord(uint64_t pointer, Word* data) PURE;
```

Comment:
This may have already been asked, but at the interface level, can't getWord/setWord just be implemented from getMemory/setMemory? Would it be better to just have utility helpers for words if that is needed?

Reply:
The API supports both 64-bit and 32-bit WASM / NullSandbox (64-bit).   So yes, it could be implemented but it would require asking the VM if it was 32-bit or 64-bit and then masking the Word.  This set of operations is so common that we added a separate API for it.   If we remove these then we have to add an API to expose whether or not the VM is 32-bit and since we had only getWord() (prior to the changes I did for htuch) it was one API or another and getWord() was the only use case for the potential isVm32Bit(). 

Do you want me to expose isVm32Bit() and remove getWord/setWord  ?   In a later PR we have a "class Wasm" which adds API support and which contains a number of helper functions (e.g. copyBuffer) and I could move getWord and setWord there.

Reply:
I'm fine either way, thanks for the explanation. I would just add more comments if you keep it here.

Reply:
Thanx. Added more comments.